subxt/src/runtime_api/runtime_payload.rs

```diff
@@ -0,0 +1,185 @@
+#[derive(Clone, Debug)]
+pub struct Payload<ArgsData, ReturnTy> {
+    fn_name: Cow<'static, str>,
+    args_data: ArgsData,
+    validation_hash: Option<[u8; 32]>,
+    _marker: PhantomData<ReturnTy>,
+}
+
+impl<ReturnTy: DecodeWithMetadata> RuntimeApiPayload for StaticRuntimeApiPayload<ReturnTy> {
+    type ReturnType = ReturnTy;
+
+    fn fn_name(&self) -> &str {
+        &self.fn_name
+    }
+
+    fn encode_args_to(&self, _metadata: &Metadata, out: &mut Vec<u8>) -> Result<(), Error> {
+        out.extend(self.args_data.clone());
+        Ok(())
+    }
+
+    fn validation_details(&self) -> Option<ValidationDetails> {
+        self.validation_hash.map(|hash| ValidationDetails { hash })
+    }
+}
+
+impl RuntimeApiPayload for DynamicRuntimeApiPayload {
+    type ReturnType = DecodedValueThunk;
+
+    fn fn_name(&self) -> &str {
+        &self.fn_name
+    }
+
+    fn encode_args_to(&self, metadata: &Metadata, out: &mut Vec<u8>) -> Result<(), Error> {
+        let fn_metadata = metadata.runtime_fn(&self.fn_name)?;
+
+        self.args_data
+            .encode_as_fields_to(fn_metadata.fields(), metadata.types(), out)?;
+
+        Ok(())
+    }
+}
+
+/// A static runtime API payload.
+///
+/// The payload already contains the scale-encoded argument bytes.
+///
+/// This is only expected to be used from codegen.
+pub type StaticRuntimeApiPayload<ReturnTy> = Payload<Vec<u8>, ReturnTy>;
+
+/// A dynamic runtime API payload.
+pub type DynamicRuntimeApiPayload = Payload<Composite<()>, DecodedValueThunk>;
+
+impl<ReturnTy: Decode, ArgsData> Payload<ArgsData, ReturnTy> {
+    /// Create a new [`Payload`].
+    pub fn new(
+        fn_name: impl Into<String>,
+        args_data: ArgsData,
+        validation_hash: Option<[u8; 32]>,
+    ) -> Self {
+        Payload {
+            fn_name: Cow::Owned(fn_name.into()),
+            args_data,
+            validation_hash,
+            _marker: PhantomData,
+        }
+    }
+
+    /// Do not validate this call prior to submitting it.
+    pub fn unvalidated(self) -> Self {
+        Self {
+            validation_hash: None,
+            ..self
+        }
+    }
+
+    /// Returns the function name.
+    pub fn fn_name(&self) -> &str {
+        &self.fn_name
+    }
+
+    /// Returns the arguments data.
+    pub fn args_data(&self) -> &ArgsData {
+        &self.args_data
+    }
+}
+
+impl<ReturnTy> StaticRuntimeApiPayload<ReturnTy> {
+    /// Create a new [`StaticRuntimeApiPayload`] using static function name
+    /// and scale-encoded argument data.
+    ///
+    /// This is only expected to be used from codegen.
+    #[doc(hidden)]
+    pub fn new_static(fn_name: &'static str, args_data: Vec<u8>, hash: [u8; 32]) -> Self {
+        Self {
+            fn_name: Cow::Borrowed(fn_name),
+            args_data,
+            validation_hash: Some(hash),
+            _marker: std::marker::PhantomData,
+        }
+    }
+}
+
+/// Create a new [`DynamicRuntimeApiPayload`].
+pub fn dynamic(
```

Comment:
it's not possible to call `DynamicRuntimeApiPaylod::new(...)` instead of this?

It looks like Payload has `fn new()`?

Reply:
This is really just for consistency I think with other modules that expose a `dynamic` function which is then made available from the `dynamic` module